types.go

```diff
@@ -1353,6 +1355,85 @@ func (rr *CSYNC) len(off int, compression map[string]struct{}) int {
 	return l
 }
 
+// APL RR. See RFC 3123.
+type APL struct {
+	Hdr      RR_Header
+	Prefixes []APLPrefix `dns:"apl"`
+}
+
+// APLPrefix is an address prefix hold by an APL record.
+type APLPrefix struct {
+	Negation bool
+	Network  net.IPNet
+}
+
+// String returns presentation form of the APL record.
+func (rr *APL) String() string {
+	sb := strings.Builder{}
+	sb.WriteString(rr.Hdr.String())
+	for i, p := range rr.Prefixes {
+		if i != 0 {
+			sb.WriteByte(' ')
+		}
+		sb.WriteString(p.String())
+	}
+	return sb.String()
+}
+
+// String returns presentation form of the APL prefix.
+func (p *APLPrefix) String() string {
+	sb := strings.Builder{}
+	if p.Negation {
+		sb.WriteByte('!')
+	}
+
+	afi := aplAddrLenToFamily[len(p.Network.IP)]
+	sb.WriteString(strconv.Itoa(int(afi)))
+
+	sb.WriteByte(':')
+
+	switch len(p.Network.IP) {
+	case net.IPv4len:
+		sb.WriteString(p.Network.IP.String())
+	case net.IPv6len:
+		// add prefix for IPv4-mapped IPv6
+		if v4 := p.Network.IP.To4(); v4 != nil {
+			sb.WriteString("::ffff:")
+		}
+		sb.WriteString(p.Network.IP.String())
+	default:
+	}
+
+	sb.WriteByte('/')
+
+	prefix, _ := p.Network.Mask.Size()
+	sb.WriteString(strconv.Itoa(prefix))
+
+	return sb.String()
+}
+
+// Equals reports whether two APL prefixes are identical.
+func (a *APLPrefix) Equals(b *APLPrefix) bool {
+	return a.Negation == b.Negation &&
+		bytes.Equal(a.Network.IP, b.Network.IP) &&
+		bytes.Equal(a.Network.Mask, b.Network.Mask)
+}
+
+// Copy returns a copy of the APL prefix.
+func (p *APLPrefix) Copy() APLPrefix {
+	return APLPrefix{
+		Negation: p.Negation,
+		Network:  copyNet(p.Network),
+	}
+}
+
+// Len returns size of the prefix in wire format.
+func (p *APLPrefix) Len() int {
```

Comment:
How useful is this method? If it's not really needed externally it should be either unexported on preferably merged into  types_generate.go.

Reply:
I'm going to make all methods on APLPrefix private. I can imagine some of them may become handy but probably not much. :+1: